### ISSUE TITLE
exception: ValueError: must be specified. connection state: "connection broken by remote peer"

### ISSUE THREAD

**Issue by vibrys** [microsoft/debugpy#1654]
Before creating a new issue, please check the [FAQ](https://github.com/microsoft/debugpy/wiki/FAQ) to see if your question is answered there.

## Environment data

- debugpy version: 1.8.5
- OS and version: Linux Ubuntu 20.04
- Python version: 3.8.10
- Using VS Code or Visual Studio: dape

## Actual behavior

exception is raised:
```
Process dape adapter stderr finished
E+00000.699: Fatal error in channel Client[1] while parsing:
             {
                 "seq": 7,
                 "type": "response",
                 "request_seq": 7,
                 "success": false,
                 "message": "Internal error",
                 "body": null
             }
             
             Traceback (most recent call last):
               File "/home/user/.pyenv/versions/mmobile/lib/python3.8/site-packages/debugpy/common/messaging.py", line 384, in __call__
                 value = validate(value)
               File "/home/user/.pyenv/versions/mmobile/lib/python3.8/site-packages/debugpy/common/json.py", line 126, in validate
                 raise ValueError("must be specified")
             ValueError: must be specified
```

## Details:
/media/user/mmobile/Scripts/python/example.py:
`print("hello world")`

jsonrpc content:
```
[jsonrpc] e[18:52:09.854] --> initialize[1] {"type":"request","seq":1,"command":"initialize","arguments":{"clientID":"dape","adapterID":"python","pathFormat":"path","linesStartAt1":true,"columnsStartAt1":true,"supportsRunInTerminalRequest":true,"supportsProgressReporting":true,"supportsStartDebuggingRequest":true}}
[jsonrpc] e[18:52:09.873] <-- initialize[1] {"seq": 4, "type": "response", "request_seq": 1, "success": true, "command": "initialize", "body": {"supportsCompletionsRequest": true, "supportsConditionalBreakpoints": true, "supportsConfigurationDoneRequest": true, "supportsDebuggerProperties": true, "supportsDelayedStackTraceLoading": true, "supportsEvaluateForHovers": true, "supportsExceptionInfoRequest": true, "supportsExceptionOptions": true, "supportsFunctionBreakpoints": true, "supportsHitConditionalBreakpoints": true, "supportsLogPoints": true, "supportsModulesRequest": true, "supportsSetExpression": true, "supportsSetVariable": true, "supportsValueFormattingOptions": true, "supportsTerminateRequest": true, "supportsGotoTargetsRequest": true, "supportsClipboardContext": true, "exceptionBreakpointFilters": [{"filter": "raised", "label": "Raised Exceptions", "default": false, "description": "Break whenever any exception is raised."}, {"filter": "uncaught", "label": "Uncaught Exceptions", "default": true, "description": "Break when the process is exiting due to unhandled exception."}, {"filter": "userUnhandled", "label": "User Uncaught Exceptions", "default": false, "description": "Break when exception escapes into library code."}], "supportsStepInTargetsRequest": true}}
[jsonrpc] e[18:52:09.873] --> launch[2] {"type":"request","seq":2,"command":"launch","arguments":{"request":"launch","type":"python","cwd":"/media/user/mmobile/Scripts/python","program":"/media/user/mmobile/Scripts/python/example.py","args":[],"justMyCode":false,"console":"integratedTerminal","showReturnValue":true,"stopOnEntry":false}}
[jsonrpc] e[18:52:09.876] <-- debugpySockets {"seq": 5, "type": "event", "event": "debugpySockets", "body": {"sockets": [{"host": "0.0.0.0", "port": 37341, "internal": false}, {"host": "127.0.0.1", "port": 57141, "internal": true}]}}
[jsonrpc] e[18:52:09.919] <-- debugpySockets {"seq": 6, "type": "event", "event": "debugpySockets", "body": {"sockets": [{"host": "0.0.0.0", "port": 37341, "internal": false}, {"host": "127.0.0.1", "port": 50049, "internal": true}, {"host": "127.0.0.1", "port": 57141, "internal": true}]}}
[jsonrpc] e[18:52:09.919] <-- runInTerminal[7] {"seq": 7, "type": "request", "command": "runInTerminal", "arguments": {"kind": "integrated", "title": "Python Debug Console", "args": ["/home/user/.pyenv/versions/mmobile/bin/python", "/home/user/.pyenv/versions/mmobile/lib/python3.8/site-packages/debugpy/launcher", "50049", "--", "/media/user/mmobile/Scripts/python/example.py"], "env": {}, "cwd": "/media/user/mmobile/Scripts/python"}}
[jsonrpc] e[18:52:10.276] --> runInTerminal[7] {"type":"response","seq":7,"request_seq":7,"success":false,"message":"Internal error","body":null}
[jsonrpc] D[18:52:10.297] Connection state change: `connection broken by remote peer
'

```

is it related to "body", which is null in seq7 response?

**Comment by rchiodo** [comment]
I would hazard a guess that DAPE isn't passing something that's expected by debugpy. You might try this with VS code and compare the messages passed (using `logToFile` in VS code launch.json)

**Comment by rchiodo** [comment]
No, I think DAPE is sending the response with an empty body. That request is from the server to the client and the response has a null body.

We follow this specification here:
https://microsoft.github.io/debug-adapter-protocol/specification#Reverse_Requests_RunInTerminal

Body is supposed to be in the response.

**Comment by svaante** [comment]
@vibrys What version of `dape` are you using? I believe this is an old issue that is fixed `M-x list-packages`

This is the response given by dape on latest (`0.14.0`)
```
[jsonrpc] e[22:19:29.761] <-- runInTerminal[7] {"seq": 7, "type": "request", "command": "runInTerminal", "arguments": {"kind": "integrated", "title": "Python Debug Console", "args": ["/Users/dpettersson/.pyenv/versions/3.8.18/bin/python", "/Users/dpettersson/.pyenv/versions/3.8.18/lib/python3.8/site-packages/debugpy/launcher", "62901", "--", "main.py"], "env": {}, "cwd": "/Users/dpettersson/Workspace/python-dap-test/"}}
[jsonrpc] e[22:19:29.777] --> runInTerminal[7] {"type":"response","seq":7,"request_seq":7,"command":"runInTerminal","success":true,"body":{"processId":7279}}
```

**Comment by vibrys** [comment]
@rchiodo, thank you for clarification.

@svaante, here's the version installed:
```
Package dape is dependency.

     Status: Installed in ‘dape-0.14.0/’.
    Version: 0.14.0
     Commit: e034ecedebc38c7f4c1e5d62044c231482b155d2
    Summary: Debug Adapter Protocol for Emacs
   Requires: emacs-29.1, jsonrpc-1.0.25
    Website: https://github.com/svaante/dape
 Maintainer: Daniel Pettersson <daniel@dpettersson.net>
Other versions: 0.14.0 (gnu).
```

your example keeps some processId. Does it mean that on my site, some process is missing at the time, the response is forged?

**Comment by vibrys** [comment]
here, I need to add the way I call `debugpy` IMO:
```
M-x dape: debugpy :cwd "/media/user/mmobile/Scripts/python" :program "/media/user/mmobile/Scripts/python/example.py"
```

maybe I'm doing some mistake around here?

**Comment by svaante** [comment]
Sorry I was to sloppy reading the logs, this is has nothing to do with returning empty body.

```
[jsonrpc] e[18:52:09.919] <-- runInTerminal[7] {"seq": 7, "type": "request", "command": "runInTerminal", "arguments": {"kind": "integrated", "title": "Python Debug Console", "args": ["/home/user/.pyenv/versions/mmobile/bin/python", "/home/user/.pyenv/versions/mmobile/lib/python3.8/site-packages/debugpy/launcher", "50049", "--", "/media/user/mmobile/Scripts/python/example.py"], "env": {}, "cwd": "/media/user/mmobile/Scripts/python"}}
[jsonrpc] e[18:52:10.276] --> runInTerminal[7] {"type":"response","seq":7,"request_seq":7,"success":false,"message":"Internal error","body":null}
```

Dape is instructed by `debugpy` to launch:
```sh
/home/user/.pyenv/versions/mmobile/bin/python /home/user/.pyenv/versions/mmobile/lib/python3.8/site-packages/debugpy/launcher 50049 -- /media/user/mmobile/Scripts/python/example.py
```
Which it fails to do and signals this by returning `"success": false`.

The python path `/home/user/.pyenv/versions/mmobile/bin/python` looks very suspicious, `.pyenv` is not an user on your machine, right?

Seams to me that  something in you env variables or `debugpy` is borked.

**Comment by vibrys** [comment]
@svaante, thank you for answer and for dape.

I'm using [pyenv](https://github.com/pyenv/pyenv) to manage virtual environments.

then my next step is to try out virtualenv(wrapper) or (maybe) even without envs, instead of pyenv.

just for completeness, more details on python path:

```
prompt$ echo $PYENV_ROOT
/home/user/.pyenv
prompt$ which python
/home/user/.pyenv/shims/python
prompt$ pyenv which python
/home/user/.pyenv/versions/mmobile/bin/python
prompt$ file /home/user/.pyenv/versions/mmobile/bin/python
/home/user/.pyenv/versions/mmobile/bin/python: symbolic link to python3
prompt$ file /home/user/.pyenv/versions/mmobile/bin/python3
/home/user/.pyenv/versions/mmobile/bin/python3: symbolic link to /usr/bin/python3
prompt$ file /usr/bin/python3
/usr/bin/python3: symbolic link to python3.8
prompt$ file /usr/bin/python3.8
/usr/bin/python3.8: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=7d4a7d9cde7d2d977b52da0a6c84bb2ee25b4217, for GNU/Linux 3.2.0, stripped
```

As an intermediate step, I'll try to play with `:python` dape/debugpy argument

**Comment by svaante** [comment]
Vibrys ***@***.***> writes:

> @svaante, thank you for answer and for dape.
No problem.

Unfortunately jsonrpc swallows the actually error, the following will
print the error to the echo area.  Hopefully that would give some more
clarity into what goes wrong (`M-x eval-defun`).

```emacs-lisp
(cl-defmethod dape-handle-request (conn (_command (eql runInTerminal)) arguments)
  "Handle runInTerminal requests.
Starts a new adapter CONNs from ARGUMENTS."
  (condition-case err
      (let ((default-directory
             (or (when-let ((cwd (plist-get arguments :cwd)))
                   (dape--path conn cwd 'local))
                 default-directory))
            (process-environment
             (or (cl-loop for (key value) on (plist-get arguments :env) by 'cddr
                          collect
                          (format "%s=%s"
                                  (substring (format "%s" key) 1)
                                  value))
                 process-environment))
            (buffer (get-buffer-create "*dape-shell*")))
        (with-current-buffer buffer
          (dape-shell-mode)
          (shell-command-save-pos-or-erase))
        (let ((process
               (make-process :name "dape shell"
                             :buffer buffer
                             :command (append (plist-get arguments :args) nil)
                             :filter 'comint-output-filter
                             :sentinel 'shell-command-sentinel
                             :file-handler t)))
          (dape--display-buffer buffer)
          (list :processId (process-id process))))
    (error
     (message "Error: %S" err))))
```

> I'm using pyenv to manage virtual environments.
> then my next step is to try out virtualenv(wrapper) or (maybe) even without envs, instead of pyenv.
I would try w/o any virtual environments to see if that works, emacs +
virtual enviroments have a rocky past (there is like 10 packages trying
to solve it).

If nothing else works you could try:
```
M-x dape: debugpy :cwd "/media/user/mmobile/Scripts/python" :program "/media/user/mmobile/Scripts/python/example.py"
:console "internalConsole"
```
Adding `:console "internalConsole"` which instructs debugpy to not send
an `runInTerminal` request.

**Comment by vibrys** [comment]
thank you for guidance

problem persists either when virtualenv(wrapper) is used or no virtualenvs at all. This in fact makes me a little bit happy - in the end virtualenvs are important part of python + your workaround with using `:console "internalConsole"` help in all 3 cases.

If I'll receive the definition of `dape-shell-mode`, then I'll provide error status from the `dape-handle-request` function (or did I miss something?)

**Comment by svaante** [comment]
Oh thats based the function from  master, if you replace dape-shell-mode with shell-mode it will be compatible with dape 0.14.0:
```emacs-lisp
(cl-defmethod dape-handle-request (conn (_command (eql runInTerminal)) arguments)
  "Handle runInTerminal requests.
Starts a new adapter CONNs from ARGUMENTS."
  (condition-case err
      (let ((default-directory
             (or (when-let ((cwd (plist-get arguments :cwd)))
                   (dape--path conn cwd 'local))
                 default-directory))
            (process-environment
             (or (cl-loop for (key value) on (plist-get arguments :env) by 'cddr
                          collect
                          (format "%s=%s"
                                  (substring (format "%s" key) 1)
                                  value))
                 process-environment))
            (buffer (get-buffer-create "*dape-shell*")))
        (with-current-buffer buffer
          (shell-mode)
          (shell-command-save-pos-or-erase))
        (let ((process
               (make-process :name "dape shell"
                             :buffer buffer
                             :command (append (plist-get arguments :args) nil)
                             :filter 'comint-output-filter
                             :sentinel 'shell-command-sentinel
                             :file-handler t)))
          (dape--display-buffer buffer)
          (list :processId (process-id process))))
    (error
     (message "Error: %S" err))))
```

**Comment by vibrys** [comment]
just changed `dape-shell-mode` to `shell-mode` and now I get following error:

```
Error: (beginning-of-buffer)
```

_*dape-repl*_ stalls on timeout:
```
* Command launch timed out after 10 seconds, the timeout period is configurable with `dape-request-timeout' *
Request timeout
* Adapter connection shutdown without successfully initializing *
Configuration:
  modes (python-mode python-ts-mode)
  ensure (lambda (config) (dape-ensure-command config) (let ((python (dape-config-get config 'command))) (unless (zerop (call-process-shell-command (format "%s -c \"import debugpy.adapter\"" python))) (user-error "%s module debugpy is not installed" python))))
  command "python"
  command-args ("-m" "debugpy.adapter" "--host" "0.0.0.0" "--port" "43193")
  port 43193
  :request "launch"
  :type "python"
  :cwd "/media/user/mmobile/Scripts/python"
  :program "/media/user/mmobile/Scripts/python/example.py"
  :args []
  :justMyCode nil
  :console "integratedTerminal"
  :showReturnValue t
  :stopOnEntry nil
  command-cwd "/media/user/mmobile/mmacs-config/public-config/"
E+00015.229: /handling #2 request "launch" from Client[1]/
             Traceback (most recent call last):
               File "/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/common/messaging.py", line 1015, in __init__
                 raise self
             debugpy.common.messaging.MessageHandlingError: Timed out waiting for launcher to connect
             
             Stack where logged:
               File "/usr/lib/python3.8/threading.py", line 890, in _bootstrap
                 self._bootstrap_inner()
               File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
                 self.run()
               File "/usr/lib/python3.8/threading.py", line 870, in run
                 self._target(*self._args, **self._kwargs)
               File "/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/common/messaging.py", line 1458, in _run_handlers
                 handler()
               File "/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/common/messaging.py", line 701, in _handle
                 result = handler(self)
               File "/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/adapter/components.py", line 91, in lock_and_handle
                 return f(self, message)
               File "/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/adapter/clients.py", line 222, in handle
                 f(self, request)
               File "/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/adapter/clients.py", line 417, in launch_request
                 launchers.spawn_debuggee(
               File "/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/adapter/launchers.py", line 168, in spawn_debuggee
                 raise start_request.cant_handle("Timed out waiting for launcher to connect")
               File "/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/common/messaging.py", line 525, in cant_handle
                 return self.error(MessageHandlingError, *args, **kwargs)
               File "/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/common/messaging.py", line 513, in error
                 exc = exc_type(reason, self, silent)  # will log it
               File "/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/common/messaging.py", line 1017, in __init__
                 log.swallow_exception()
               File "/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/common/log.py", line 215, in swallow_exception
                 _exception(format_string, *args, **kwargs)
             

E+00015.242: /handling #2 request "launch" from Client[1]/
             Handler 'Client._start_message_handler.<locals>.handle' (file '/media/user/mmobile/.local/lib/python3.8/site-packages/debugpy/adapter/components.py', line 206)
             couldn't handle #2 request "launch" from Client[1]:
             Timed out waiting for launcher to connect

```

it might be that timeout is due to `dape-handle-request` is not succeeding, right?

the test was done with NO virtualenvs enabled
